vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -212,23 +212,25 @@
         if (isEmpty()) {
             return Option.none();
         } else {
```

Comment:
this `else` is unnecessary. Once you have done `return Option.none();` the only way the rest of the code executes is as the `else`. Remove the else -> remove indent -> reduce complexity

Reply:
There was discussion about 'guards' before. We decided to

* Do method parameter checks as guards (, like you suggested)
* Do control flow of the method body the 'functional way' as if the `if` statement where an expression, like in java. Logically the two cases belong to each other. Flattening a method is the imperative way. Some information gets lost. I know that some say the the complexity increases. For me, the readability decreases.

Example of your suggestions below:

<img width="927" alt="screen shot 2017-09-12 at 20 02 48" src="https://user-images.githubusercontent.com/743833/30341346-010d9120-97f6-11e7-94e2-70beaeeb6362.png">

Reply:
Imperative isn't a crime OR a dirty word. I noticed there is a cyclomatic complexity analyzer on the build. Reducing that number seems an unalloyed improvement.

Reply:
Thank you for your feedback! Sometimes keeping the order is more important than a number. E.g. imagine we flatten two scoped blocks to one. No-one hinders someone to change the order of the statements later.

Example:

```java
if (condition1) {
    do();
    this();
    and();
    that();
} else {
    throw x;
}
if (condition2) {
    flip();
    flap();
} else {
    throw y;
}
```

The following might be a valid substitute:

```java
if (!condition1) {
    throw x;
}
if (!condition2) {
    throw y;
}
do();
flip();
this();
and();
flap();
that();
```

But some important information got lost. Do you see what I mean?

The cyclomatic complexity analyzer is 'stupid'. It can't look through humans eyes.

Reply:
@chb0github you points are valid. I made some tests and looked in the java.util.stream.DoubleStream.average() algorithm. It is accurate, I think we do not need to convert elements to long.

This leads to the following impl:

```java
    default Option<Double> average() {
        if (isEmpty()) {
            return Option.none();
        }
        try {
            final java.util.stream.Stream<Number> numbers = (java.util.stream.Stream<Number>) toJavaStream();
            final double avg = numbers.mapToDouble(Number::doubleValue).average().getAsDouble();
            return Option.some(avg);
        } catch (ClassCastException x) {
            throw new UnsupportedOperationException("not numeric");
        }
    }
```

Reply:
Honestly, toss the catch statement. You're trading one exception for another and the `ClassCastException` says everything you need to know

Reply:
Your conditional collapse substitution is not identical to the original sample.

This would be the equivalence:

```
if (condition1) {
    do;
    this;
    and;
    that;
} else {
    throw x;
}
if (condition2) {
    flip;
    flap;
} else {
    throw y;
}
```

```
if (!condition1) {
   throw x;
}
do;
this;
and;
that;

if (!condition2) {
    throw y;
}
flip;
flap;
```

And, actually, I don't see what got lost except extraneous syntax.

Reply:
ClassCastEcxeption is an implementation detail. It might change.

Reply:
The rule we follow:

1. If we check an argument we do a guard (most often Objects.requireNonNull(), but also if-guards)
2. After checking the arguments, we use if-else

```java
// THE WAY WE DO IT
default Option<T> maxBy(Comparator<? super T> comparator) {
    Objects.requireNonNull(comparator, "comparator is null");
    if (isEmpty()) {
        return Option.none();
    } else {
        final T value = reduce((t1, t2) -> comparator.compare(t1, t2) >= 0 ? t1 : t2);
        return Option.some(value);
    }
}

// NOT ACCORDING TO OUR CODING GUIDELINE
default Option<T> maxBy(Comparator<? super T> comparator) {
    Objects.requireNonNull(comparator, "comparator is null");
    if (isEmpty()) {
        return Option.none();
    }
    final T value = reduce((t1, t2) -> comparator.compare(t1, t2) >= 0 ? t1 : t2);
    return Option.some(value);
}
```

We justify the guideline mentioned above by keeping things together with if-else which logically belog together. I.e. the two-sided conditional is a composition of a fork in our code.

If we have more than one of these logical units (read: if-else) we also benefit from scoping of variables that are needed only within the branches. Otherwise we would expose these variables to contexts which should not know about them.

Summing up, if-else helps us to structure our code and making it more robust.